pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1808,43 +1808,68 @@ public PinotResourceManagerResponse disableInstance(String instanceName) {
    * @return
    */
-  public PinotResourceManagerResponse dropInstance(String instanceName) {
-    if (!instanceExists(instanceName)) {
-      return new PinotResourceManagerResponse("Instance " + instanceName + " does not exist.", false);
-    }
-
+  public boolean isInstanceDroppable(String instanceName) {
+    // Check if this instance is live
     HelixDataAccessor helixDataAccessor = _helixZkManager.getHelixDataAccessor();
     LiveInstance liveInstance = helixDataAccessor.getProperty(_keyBuilder.liveInstance(instanceName));
-
     if (liveInstance != null) {
-      PropertyKey currentStatesKey =
-          _keyBuilder.currentStates(instanceName, liveInstance.getSessionId());
-      List<CurrentState> currentStates = _helixDataAccessor.getChildValues(currentStatesKey);
+      return false;
+    }
 
-      if (currentStates != null) {
-        for (CurrentState currentState : currentStates) {
-          for (String state : currentState.getPartitionStateMap().values()) {
-            if (state.equalsIgnoreCase(SegmentOnlineOfflineStateModel.ONLINE)) {
-              return new PinotResourceManagerResponse(("Instance " + instanceName + " has online partitions"), false);
-            }
+    // Check if any idealstate contains information on this instance
+    for (String tableName : getAllTableNames()) {
+      IdealState tableIdealState = _helixAdmin.getResourceIdealState(_helixClusterName, tableName);
+      for (String partition : tableIdealState.getPartitionSet()) {
+        for (String instance : tableIdealState.getInstanceSet(partition)) {
+          if (instance.equals(instanceName)) {
+            return false;
           }
         }
-      } else {
-        return new PinotResourceManagerResponse("Cannot drop live instance " + instanceName
-            + " please stop the instance first.", false);
       }
     }
+    return true;
+  }
 
-    // Disable the instance first.
-    toogleInstance(instanceName, false, 10);
-    _helixAdmin.dropInstance(_helixClusterName, getHelixInstanceConfig(instanceName));
+  /**
+   * Drop the instance from helix cluster. Instance will not be dropped if:
+   *
+   * @param instanceName: Name of the instance to be dropped.
+   * @return
+   */
+  public PinotResourceManagerResponse dropInstance(final String instanceName) {
+    // Delete '/CONFIGS/PARTICIPANT/<server_name>'
+    boolean successfulConfigRemove = DEFAULT_RETRY_POLICY.attempt(new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        PropertyKey instanceKey = _keyBuilder.instanceConfig(instanceName);
+        return _helixDataAccessor.removeProperty(instanceKey);
+      }
+    });
+    if (!successfulConfigRemove) {
+      return new PinotResourceManagerResponse("fail to erase /CONFIGS/PARTICIPANT/<server_name> for instance "
+          + instanceName, false);
+    }
+
+    // Delete '/INSTANCES/<server_name>'
+    final String instancePath = "/" + _helixClusterName + "/INSTANCES/" + instanceName;
```

Comment:
Do these steps in the reverse order (delete an instance, then its configuration). Otherwise, it's possible to find instances that have no configuration, which caused us issues with Helix.